### PR TITLE
fix: thread safety regarding env vars

### PR DESCRIPTION
**Pull request by shouze**
Fix: #632

### CODE REVIEW

File: src/Core/src/Credentials/IniFileLoader.php [hunk 1]
```diff
@@ -65,13 +65,13 @@ public function loadProfiles(array $filepaths): array
     private function getHomeDir(): string
     {
         // On Linux/Unix-like systems, use the HOME environment variable
-        if (false !== $homeDir = \getenv('HOME')) {
+        if (false !== $homeDir = $_SERVER['HOME']) {
```
Comment (nicolas-grekas) [inline src/Core/src/Credentials/IniFileLoader.php:68]: `if (\is_string($homeDir = $_SERVER['HOME'] ?? null)) {`

File: src/Core/src/Credentials/IniFileLoader.php [hunk 2]
```diff
@@ -65,13 +65,13 @@ public function loadProfiles(array $filepaths): array
...
             return $homeDir;
         }
 
         // Get the HOMEDRIVE and HOMEPATH values for Windows hosts
-        $homeDrive = \getenv('HOMEDRIVE');
-        $homePath = \getenv('HOMEPATH');
+        $homeDrive = $_SERVER['HOMEDRIVE'];
```
Comment (nicolas-grekas) [inline src/Core/src/Credentials/IniFileLoader.php:73]: `?? null` (same below)

File: src/Core/src/Configuration.php [hunk 3]
```diff
@@ -93,8 +93,8 @@ public static function create(array $options)
             foreach ($fallbackGroup as $option => $envVariableNames) {
                 $envVariableNames = (array) $envVariableNames;
                 foreach ($envVariableNames as $envVariableName) {
-                    if (false !== $value = \getenv($envVariableName)) {
-                        $options[$option] = $value;
+                    if (\array_key_exists($envVariableName, $_SERVER)) {
```
Comment (nicolas-grekas) [inline src/Core/src/Configuration.php:96]: `isset()` is enough - `null` is not a legit value anyway

**Comment by Nyholm** [comment]
Will Lambda/EC2 populate the `$_SERVER` variable with the environment variables?

**Comment by chalasr** [comment]
@Nyholm Yes, no issue here (belongs to the PHP runtime). See e.g. https://bref.sh/docs/runtimes/http.html#context-access

**Comment by Nyholm** [comment]
That is exactly what Im worried about. This is perfectly fine to run on Lambda (bref) and Ec2: 

```php
$s3 = new S3Client();
$result = $s3->getObject(['Bucket'=>'my-private-bucket', 'Key'=>'my-file.jpg');
```

The authentication credentials is in the environment variables. 

However, after merging this PR, will the same code be able to run? Im not sure that bref is setting `$_SERVER['AWS_ACCESS_KEY_ID ']` and `$_SERVER['AWS_SECRET_ACCESS_KEY ']`. If they are not, this will PR will break their code in production runtime. 

Same thing with EC2. Is `$_SERVER['AWS_ACCESS_KEY_ID ']` available when you running CLI commands?

**Comment by jderusse** [comment]
I don't know how bref works, but env variables are populated into `$_SERVER`. This PR will not be an issue for CLI command on EC2.

**Comment by shouze** [comment]
@Nyholm yes for sure it will be populated into bref.

FYI, I've squashed my 2 commits into a single one I think we're ready for a merge.

**Comment by Nyholm** [comment]
You are all correct. I had to test myself just to be sure. 

Thank you for the PR and thank you for all the reviews.

**Comment by mnapoli** [comment]
This is so weird, this broke in my local environment (running Symfony on my dev machine with `php -S 127.0.0.1:8000 -t public`).

After debugging for an hour a failed presigned S3 URL, I realized that:

```php
getenv('HOME'); // works
$_SERVER['HOME']; // undefined
```

which broke the authentication of the SDK.

Now I remember I had to explicitly enable the environment variables in `$_SERVER` in Bref (https://github.com/brefphp/bref/blob/master/runtime/layers/fpm/php.ini#L32). I'm surprised to see that PHP's installation on MacOS with Brew does not enable it:

```ini
variables_order = "GPCS"
```

Maybe the problem is with my laptop (in which case I don't know why), but is it?

- are environment variables populated for you (`php -i | grep variables_order`) on your laptop?
- in general, if PHP recommends `GPCS` as the good value for production and development, should we still use this?

As you can probably guess I'm full of questions and have no idea if there's a problem at all here :/

**Comment by shouze** [comment]
> * php -i | grep variables_order

I'm using brew under Linux and here's what I have:
```shell
❯ php -i | grep variables_order
variables_order => GPCS => GPCS
```
So far no env vars loaded too. Does launching php test server work this way?
```shell
php -c variables_order=EGPCS -S 127.0.0.1:8000 -t
``

**Comment by mnapoli** [comment]
Yes it should work (`php -c variables_order=EGPCS -i | grep variables_order` does work), but does that mean that we should revert the PR? Because it might be broken for more people then?

**Comment by shouze** [comment]
> Yes it should work (`php -c variables_order=EGPCS -i | grep variables_order` does work), but does that mean that we should revert the PR? Because it might be broken for more people then?

I'm not sure it's a good idea to revert now. Thread safety if far more important for production than being able to use php test server. I understand the frustraction/bad DX of course so it would be nice to find a way to improve DX without compromising production safety. I have no idea at the moment but I think it was important to get agree on that first.

**Comment by jderusse** [comment]
what's about fallback to gentenv when value in `$_SERVER` is null?

or use an internal `env` method that check the value of ini setting to return either `$_server` or `getenv`.

**Comment by mnapoli** [comment]
> Thread safety if far more important for production than being able to use php test server.

I'm not sure it limits to "php test server". It applies to PHP installed with brew on Mac and Linux: the PHP cli, php-fpm and `php -S`. It impacts a lot of people.

Furthermore, this is the consequence of _the default PHP config_. So it's perfectly possible it's affecting production environments too.

Actually, I had a look at the `php.ini` shipped by PHP itself:

development version: https://github.com/php/php-src/blob/56135654a0f2751068cd27317f7dab9f483ff456/php.ini-development#L637-L648
production version: https://github.com/php/php-src/blob/56135654a0f2751068cd27317f7dab9f483ff456/php.ini-production#L637-L648

In both cases, the default value is `GPCS`, and the recommended values are `GPCS`, meaning this PR does not work with these settings.

**Comment by Nyholm** [comment]
Im on OSX with PHP installed from brew: 

```
php -i | grep variables_order
variables_order => GPCS => GPCS
```

My coworker in on Ubuntu 20.04.1:

```
php -i | grep variables_order
variables_order => GPCS => GPCS
```

We both use the "stock" configuration for this value. 

----------

To clarify: 
1. Are you saying it doesn't work on `GPCS`? 
2. Does it work with `EGPCS`?

**Comment by acsfer** [comment]
@mnapoli we are calling `$_SERVER` here, not `$_ENV`, so **GPCS** defaults should satisfy this.

See also https://github.com/symfony/symfony/issues/26451#issuecomment-373153726

**Comment by mnapoli** [comment]
Wait no, I did more testing!

>    Are you saying it doesn't work on GPCS?

It doesn't work with GPCS.

>    Does it work with EGPCS?

It doesn't work with EGPCS either, I read the config too fast. `E` is for `$_ENV`, not "populating environment variables" in `$_SERVER`.

However, after testing some more, including with the official PHP docker image, here is what I see:

- GPCS or EGPCS doesn't matter
- `$_SERVER['xxx']` works in CLI
- `$_SERVER['xxx']` **doesn't work with PHP's built-in server**

So you were right @shouze, the scope seems to be more limited than I thought.

Here is how you can reproduce (works also without docker):

```
docker run --rm -it -p 8000:8000 --env TEST=abc -v $PWD:/app -w /app php -S 0.0.0.0:8000
```

Create a `index.php` file with:

```php
<?php

echo getenv('TEST');
echo $_SERVER['TEST'];
```

Only the first one will be populated. I'm not sure why this happening, could it be a bug in the built-in server?

**Comment by acsfer** [comment]
See the comment on my post ;)

**Comment by Nyholm** [comment]
I did some testing:  (All with "stock" php config)

| Server | `getenv` | `$_SERVER` |
| ------ | --------- | ------------ |
| php -S | Yes | No |
| php -S (with docker) | Yes | No |
| Symfony webserver | Yes | Yes |
| php-fpm Nginx | Yes | Yes |
| php-fpm on Lambda with Bref | Yes | Yes |

It might be something wrong with the built-in PHP server.

**Comment by jderusse** [comment]
#769 should fix this issue.
/cc @mnapoli

**Pull request merged**